utils/env_preferences/src/apple.rs

```diff
@@ -0,0 +1,136 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+pub mod apple_prefs {
+    use core_foundation_sys::{
+        array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+        base::{CFRelease, CFRetain},
+        calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+        locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+        string::CFStringGetCStringPtr,
+    };
+    use std::ffi::CStr;
+
+    pub fn get_locales() -> Vec<String> {
+        let mut languages: Vec<String> = Vec::new();
+
+        // SAFETY: The call to `CFLocaleCopyPreferredLanguages` returns an immutable reference to `CFArray` which is owned by us
+        // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
+        // Immutablility ensures that nothing is overriden during it's scope
+        let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
+
+        if !locale_carr_ref.is_null() {
+            // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
+            let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
```

Comment:
at this point you can be optimistic and call `locales.reserve(count)` to avoid reallocating in the loop.